### PR TITLE
CORS-2933: IBMCloud: Basic service endpoint override

### PR DESCRIPTION
**Pull request by cjschaef**
Adding basic support to allow overrides of IBM Cloud service endpoints. This is a hard requirement for disconnected cluster support, to allow use of non-default endpoints for IBM Cloud services.

### CODE REVIEW

File: pkg/asset/installconfig/ibmcloud/validation.go [hunk 1]
```diff
@@ -414,6 +417,54 @@ func ValidatePreExistingPublicDNS(client API, ic *types.InstallConfig, metadata
...
+
+		// Check if the service URL is valid
+		err := validateEndpoint(service.URL)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(serviceEndpointsPath.Index(id).Child("url"), service.URL, err.Error()))
+			continue
+		}
+	}
+
+	// TODO(cjschaef): See about performing additional endpoint validation (all vs. subsets, regional endpoint validation, etc.)
```
Comment (rvanderp3) [inline pkg/asset/installconfig/ibmcloud/validation.go:448]: what is the status of this TODO?

File: pkg/asset/machines/ibmcloud/zones.go [hunk 2]
```diff
@@ -3,14 +3,15 @@ package ibmcloud
 import (
 	"context"
 
+	configv1 "github.com/openshift/api/config/v1"
 	"github.com/openshift/installer/pkg/asset/installconfig/ibmcloud"
```
Comment (rvanderp3) [inline pkg/asset/machines/ibmcloud/zones.go:7]:
```suggestion

	"github.com/openshift/installer/pkg/asset/installconfig/ibmcloud"
```

File: pkg/asset/manifests/ibmcloud/cloudproviderconfig.go [hunk 3]
```diff
@@ -3,9 +3,13 @@ package ibmcloud
 import (
 	"bytes"
 	"fmt"
+	"regexp"
 	"sort"
 	"strings"
 	"text/template"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/installer/pkg/types/ibmcloud"
```
Comment (rvanderp3) [inline pkg/asset/manifests/ibmcloud/cloudproviderconfig.go:12]:
```suggestion
	configv1 "github.com/openshift/api/config/v1"
	
	"github.com/openshift/installer/pkg/types/ibmcloud"
```

File: pkg/asset/manifests/ibmcloud/cloudproviderconfig.go [hunk 4]
```diff
@@ -67,6 +74,21 @@ func CloudProviderConfig(infraID string, accountID string, region string, resour
...
+		switch name := strings.ToLower(endpoint.Name); name {
+		case ibmcloud.IBMCloudServiceIAM:
+			config.Provider.IAMEndpointOverride = endpoint.URL
+		case ibmcloud.IBMCloudServiceVPC:
+			// Trim the version suffix ('/v1', '/v21', etc.) from the VPC URL since the CCM appends it
+			suffixRegex := regexp.MustCompile(`(^.*)/v\d+[/]{0,1}$`)
+			config.Provider.VPCEndpointOverride = suffixRegex.ReplaceAllString(endpoint.URL, "${1}")
+		case ibmcloud.IBMCloudServiceResourceManager:
+			config.Provider.RMEndpointOverride = endpoint.URL
+		}
```
Comment (rvanderp3) [inline pkg/asset/manifests/ibmcloud/cloudproviderconfig.go:89]: would there be any value in logging when this switch falls through?

File: pkg/destroy/ibmcloud/ibmcloud.go [hunk 5]
```diff
@@ -23,9 +23,11 @@ import (
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/apimachinery/pkg/util/wait"
 
+	configv1 "github.com/openshift/api/config/v1"
 	icibmcloud "github.com/openshift/installer/pkg/asset/installconfig/ibmcloud"
 	"github.com/openshift/installer/pkg/destroy/providers"
 	"github.com/openshift/installer/pkg/types"
+	ibmcloudtypes "github.com/openshift/installer/pkg/types/ibmcloud"
 	"github.com/openshift/installer/pkg/version"
```
Comment (rvanderp3) [inline pkg/destroy/ibmcloud/ibmcloud.go:31]:
```suggestion
	configv1 "github.com/openshift/api/config/v1"
	
	icibmcloud "github.com/openshift/installer/pkg/asset/installconfig/ibmcloud"
	"github.com/openshift/installer/pkg/destroy/providers"
	"github.com/openshift/installer/pkg/types"
	ibmcloudtypes "github.com/openshift/installer/pkg/types/ibmcloud"
	"github.com/openshift/installer/pkg/version"
```

File: pkg/tfvars/ibmcloud/ibmcloud.go [hunk 6]
```diff
@@ -2,14 +2,21 @@ package ibmcloud
...
+	"strings"
 
 	"github.com/pkg/errors"
 
+	configv1 "github.com/openshift/api/config/v1"
 	"github.com/openshift/installer/pkg/tfvars/internal/cache"
 	"github.com/openshift/installer/pkg/types"
+	ibmcloudtypes "github.com/openshift/installer/pkg/types/ibmcloud"
 	ibmcloudprovider "github.com/openshift/machine-api-provider-ibmcloud/pkg/apis/ibmcloudprovider/v1"
 )
```
Comment (rvanderp3) [inline pkg/tfvars/ibmcloud/ibmcloud.go:15]:
```suggestion
	configv1 "github.com/openshift/api/config/v1"
	
	"github.com/openshift/installer/pkg/tfvars/internal/cache"
	"github.com/openshift/installer/pkg/types"
	ibmcloudtypes "github.com/openshift/installer/pkg/types/ibmcloud"
	ibmcloudprovider "github.com/openshift/machine-api-provider-ibmcloud/pkg/apis/ibmcloudprovider/v1"
)
```

File: pkg/tfvars/ibmcloud/ibmcloud.go [hunk 7]
```diff
@@ -96,26 +105,56 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
...
+// CreateEndpointJSON creates JSON data containing IBM Cloud service endpoint override mappings.
+func CreateEndpointJSON(endpoints []configv1.IBMCloudServiceEndpoint, region string) ([]byte, error) {
+	// If no endpoint overrides, simply return
+	if len(endpoints) == 0 {
+		return nil, nil
+	}
+
+	// Following the format provided by the IBM Cloud Terraform Provider documentation:
+	// https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/guides/custom-service-endpoints#file-structure-for-endpoints-file
+	endpointContents := make(map[string]map[string]map[string]string)
```
Comment (rvanderp3) [inline pkg/tfvars/ibmcloud/ibmcloud.go:142]: this seems like a fairly complex way of representing this data structure.  would it be possible to create a type representing it?

File: pkg/types/ibmcloud/validation/platform.go [hunk 8]
```diff
@@ -1,8 +1,13 @@
...
 
 import (
+	"fmt"
+	"net/url"
+	"regexp"
+
 	"k8s.io/apimachinery/pkg/util/validation/field"
 
+	configv1 "github.com/openshift/api/config/v1"
 	"github.com/openshift/installer/pkg/types/ibmcloud"
```
Comment (rvanderp3) [inline pkg/types/ibmcloud/validation/platform.go:11]:
```suggestion
	"k8s.io/apimachinery/pkg/util/validation/field"
	configv1 "github.com/openshift/api/config/v1"
	
	"github.com/openshift/installer/pkg/types/ibmcloud"
```

File: pkg/types/ibmcloud/validation/platform.go [hunk 9]
```diff
@@ -58,5 +63,68 @@ func ValidatePlatform(p *ibmcloud.Platform, fldPath *field.Path) field.ErrorList
...
+// schemeRE is used to check whether a string starts with a scheme (URI format).
+var schemeRE = regexp.MustCompile("^([^:]+)://")
+
+// validateServiceURL checks that a string meets certain URI expectations.
+func validateServiceURL(uri string) error {
+	endpoint := uri
+	httpsScheme := "https"
+
+	// Regexp for a trailing API version in URL path ('/v1', '/v22/', etc.)
+	versionPath := regexp.MustCompile(`(/v\d+[/]{0,1})$`)
```
Comment (rvanderp3) [inline pkg/types/ibmcloud/validation/platform.go:101]: would it make sense to compile this regex once as in line 93?

File: pkg/types/ibmcloud/validation/platform_test.go [hunk 10]
```diff
@@ -6,6 +6,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 
+	configv1 "github.com/openshift/api/config/v1"
 	"github.com/openshift/installer/pkg/types/ibmcloud"
```
Comment (rvanderp3) [inline pkg/types/ibmcloud/validation/platform_test.go:10]:
```suggestion
	"k8s.io/apimachinery/pkg/util/validation/field"
	configv1 "github.com/openshift/api/config/v1"
	
	"github.com/openshift/installer/pkg/types/ibmcloud"
```

**Review by rvanderp3 — COMMENTED** [review]
overall LGTM

File: pkg/tfvars/ibmcloud/ibmcloud.go (hunk 7 above)
Comment (cjschaef) [reply pkg/tfvars/ibmcloud/ibmcloud.go:142]: Sure, I'll see about doing that instead, to get the JSON format correct.

File: pkg/asset/installconfig/ibmcloud/validation.go (hunk 1 above)
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/validation.go:448]: I was expecting to follow up on this later, as I wait for details from other development teams on possible configurations, environments, etc. that would determine if we wish to perform additional validation, or leave the endpoint override possibilities broader.

File: pkg/asset/manifests/ibmcloud/cloudproviderconfig.go (hunk 4 above)
Comment (cjschaef) [reply pkg/asset/manifests/ibmcloud/cloudproviderconfig.go:89]: Maybe. Other endpoints the CCM doesn't use (COS, DNS Services, etc.), would fall through, so we'd ignore them in terms of building the CCM config.
But perhaps logging the "ignored" overrides could be helpful if the user has incorrectly configured those required for the CCM, but I'd expect the `validation` steps should filter those issues out
https://github.com/openshift/installer/pull/7632/files#diff-e6306df593d77282680d1864f5e1504e044abdbe681195fbcbeaab0cc3c45cc2R437

I'll add a `default` log call I think though.

File: pkg/types/ibmcloud/validation/platform.go (hunk 9 above)
Comment (cjschaef) [reply pkg/types/ibmcloud/validation/platform.go:101]: Oh yeah, I can move that.

File: pkg/asset/manifests/ibmcloud/cloudproviderconfig.go (hunk 4 above)
Comment (cjschaef) [reply pkg/asset/manifests/ibmcloud/cloudproviderconfig.go:93]: Added a log call

File: pkg/tfvars/ibmcloud/ibmcloud.go (hunk 7 above)
Comment (cjschaef) [reply pkg/tfvars/ibmcloud/ibmcloud.go]: I made the design a bit better I hope.

File: pkg/types/ibmcloud/validation/platform.go (hunk 9 above)
Comment (cjschaef) [reply pkg/types/ibmcloud/validation/platform.go]: Moved

**Comment by cjschaef** [comment]
>Init container clonerefs not ready: (state: running) Init container initupload not ready: (state: waiting, reason: "PodInitializing", message: "") Init container place-entrypoint not ready: (state: waiting, reason: "PodInitializing", message: "")

/retest

File: pkg/asset/machines/ibmcloud/zones.go (hunk 2 above)
Comment (cjschaef) [reply pkg/asset/machines/ibmcloud/zones.go:8]: I believe these imports are formatted by `gofmt`, and making these changes results in that test failing
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/7632/pull-ci-openshift-installer-master-gofmt/1721621569781895168

For example:
```
diff --git a/pkg/asset/machines/ibmcloud/zones.go b/pkg/asset/machines/ibmcloud/zones.go
index fe0fd4323f..d222b3a8dc 100644
--- a/pkg/asset/machines/ibmcloud/zones.go
+++ b/pkg/asset/machines/ibmcloud/zones.go
@@ -4,7 +4,6 @@ import (
 	"context"
 
 	configv1 "github.com/openshift/api/config/v1"
-
 	"github.com/openshift/installer/pkg/asset/installconfig/ibmcloud"
 )
```

Comment (cjschaef) [reply pkg/asset/machines/ibmcloud/zones.go:8]: I'm going to resolve the `gofmt` items, but think that will result in the same imports as before attempting to address your comments in that regard.

File: pkg/types/ibmcloud/validation/platform_test.go (hunk 10 above)
Comment (cjschaef) [reply pkg/types/ibmcloud/validation/platform_test.go:10]: `gofmt` preferred this as is, so I've changed it back
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/7632/pull-ci-openshift-installer-master-gofmt/1721621569781895168

File: pkg/types/ibmcloud/validation/platform.go (hunk 8 above)
Comment (cjschaef) [reply pkg/types/ibmcloud/validation/platform.go:11]: `gofmt` preferred this as is, so I've changed it back
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/7632/pull-ci-openshift-installer-master-gofmt/1721621569781895168

File: pkg/tfvars/ibmcloud/ibmcloud.go (hunk 6 above)
Comment (cjschaef) [reply pkg/tfvars/ibmcloud/ibmcloud.go:15]: `gofmt` preferred this as is, so I've changed it back
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_installer/7632/pull-ci-openshift-installer-master-gofmt/1721621569781895168

**Comment by cjschaef** [comment]
/retitle CORS-2934: IBMCloud: Basic service endpoint override

File: pkg/tfvars/ibmcloud/ibmcloud.go (hunk 6 above)
Comment (rvanderp3) [reply pkg/tfvars/ibmcloud/ibmcloud.go:15]: makes sense, sorry about that.

File: pkg/asset/installconfig/ibmcloud/validation.go (hunk 1 above)
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/validation.go]: I removed the TODO, as we will be leaving the configuration options more broad, due to the variety of environments/restrictions we can cover.

File: pkg/tfvars/ibmcloud/ibmcloud.go (hunk 6 above)
Comment (cjschaef) [reply pkg/tfvars/ibmcloud/ibmcloud.go:15]: No worries, I always have trouble with them without `gofmt`'s help.

**Comment by cjschaef** [comment]
I am updating to pull in the changes from `api`, for service names.
https://github.com/openshift/api/commit/2155ee0c44e925e3642db5781cd4a4e2f1808376

**Comment by cjschaef** [comment]
/retitle CORS-2933: IBMCloud: Basic service endpoint override

**Comment by r4f4** [comment]
/hold
This depends on https://github.com/openshift/installer/pull/7617 for the `openstack-manifests` test fix.

**Comment by cjschaef** [comment]
I'll rebase off https://github.com/openshift/installer/pull/7617, once it merges, as we will still need a newer `api` release.

File: pkg/destroy/bootstrap/bootstrap.go [hunk 11]
```diff
@@ -42,6 +47,27 @@ func Destroy(dir string) (err error) {
 		platform = typesazure.StackTerraformName
 	}
 
+	// IBM Cloud allows override of service endpoints, which would be required during bootstrap destroy.
+	// Create a JSON file with overrides, if these endpoints are present
+	if platform == ibmcloudtypes.Name && metadata.IBMCloud != nil && len(metadata.IBMCloud.ServiceEndpoints) > 0 {
+		// Build the JSON containing the endpoint overrides for IBM Cloud Services.
+		jsonData, err := ibmcloudtfvars.CreateEndpointJSON(metadata.IBMCloud.ServiceEndpoints, metadata.IBMCloud.Region)
+		if err != nil {
+			return errors.Wrap(err, "failed generating endpoint override JSON data for bootstrap destroy")
```
Comment (r4f4) [inline pkg/destroy/bootstrap/bootstrap.go:56]:
```suggestion
			return fmt.Errorf("failed generating endpoint override JSON data for bootstrap destroy: %w", err)
```

File: pkg/destroy/bootstrap/bootstrap.go [hunk 12]
```diff
@@ -42,6 +47,27 @@ func Destroy(dir string) (err error) {
...
+		}
+		// Since there are ServiceEndpoints, we expect JSON data to be generated.
+		if jsonData == nil {
+			return errors.New("no endpoint override JSON data generated for set of endpoint overrides")
+		}
+
+		// If JSON data was generated, create the JSON file for IBM Cloud Terraform provider to use during destroy.
+		endpointsFilePath := filepath.Join(dir, ibmcloudtfvars.IBMCloudEndpointJSONFileName)
+		if err := os.WriteFile(endpointsFilePath, jsonData, 0o600); err != nil {
+			return errors.Wrap(err, "failed to write IBM Cloud service endpoint override JSON file for bootstrap destroy")
```
Comment (r4f4) [inline pkg/destroy/bootstrap/bootstrap.go:66]:
```suggestion
			return fmt.Errorf("failed to write IBM Cloud service endpoint override JSON file for bootstrap destroy: %w", err)
```

File: pkg/terraform/stages/ibmcloud/stages.go [hunk 13]
```diff
@@ -17,11 +27,29 @@ var PlatformStages = []terraform.Stage{
...
+	// If these is a endpoint override JSON file in the terraformDir's parent directory (terraformDir isn't available during JSON file creation),
+	// we want to inject that file into the Terraform variables so IBM Cloud Service endpoints are overridden.
+	terraformParentDir := filepath.Dir(terraformDir)
+	endpointOverrideFile := filepath.Join(terraformParentDir, ibmcloudtfvars.IBMCloudEndpointJSONFileName)
+	if _, err := os.Stat(endpointOverrideFile); err == nil {
+		// Set variable to use private endpoints (overrides) from JSON file, via the IBM Cloud Terraform variable: 'ibmcloud_endpoints_json_file'.
+		opts = append(opts, tfexec.Var(fmt.Sprintf("ibmcloud_endpoints_json_file=%s", endpointOverrideFile)))
+		logrus.Debugf("configuring terraform bootstrap destroy with ibm endpoint overrides: %s", endpointOverrideFile)
+	}
+	return errors.Wrap(terraform.Destroy(directory, ibmcloudtypes.Name, s, terraformDir, opts...), "failed to destroy bootstrap")
```
Comment (r4f4) [inline pkg/terraform/stages/ibmcloud/stages.go:54]:
```suggestion
   err := terraform.Destroy(directory, ibmcloudtypes.Name, s, terraformDir, opts...)
   if err != nil {
       return fmt.Errorf("failed to destroy bootstrap: %w", err)
   }
   return nil
```

File: pkg/tfvars/ibmcloud/ibmcloud.go [hunk 14]
```diff
@@ -96,26 +104,119 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
...
+					region: endpoint.URL,
+				},
+			}
+		default:
+			return nil, fmt.Errorf("unable to build override values for unknown service: %s", endpoint.Name)
+		}
+	}
+	jsonData, err := json.Marshal(endpointContents)
+	if err != nil {
+		return nil, errors.Wrap(err, "failure building service endpoint override JSON data")
```
Comment (r4f4) [inline pkg/tfvars/ibmcloud/ibmcloud.go:214]:
```suggestion
		return nil, fmt.Errorf("failure building service endpoint override JSON data: %w", err)
```

Comment (cjschaef) [reply pkg/tfvars/ibmcloud/ibmcloud.go:214]: I'll migrate to `fmt.Errorf`

File: pkg/terraform/stages/ibmcloud/stages.go (hunk 13 above)
Comment (cjschaef) [reply pkg/terraform/stages/ibmcloud/stages.go:54]: I'll migrate to `fmt.Errorf`

File: pkg/destroy/bootstrap/bootstrap.go (hunk 12 above)
Comment (cjschaef) [reply pkg/destroy/bootstrap/bootstrap.go:66]: I'll migrate to `fmt.Errorf`

File: pkg/destroy/bootstrap/bootstrap.go (hunk 11 above)
Comment (cjschaef) [reply pkg/destroy/bootstrap/bootstrap.go:56]: I'll migrate to `fmt.Errorf`

File: pkg/asset/installconfig/ibmcloud/client.go [hunk 15]
```diff
@@ -460,7 +516,7 @@ func (c *Client) GetVSIProfiles(ctx context.Context) ([]vpcv1.InstanceProfile, e
 	listInstanceProfilesOptions := c.vpcAPI.NewListInstanceProfilesOptions()
 	profiles, _, err := c.vpcAPI.ListInstanceProfilesWithContext(ctx, listInstanceProfilesOptions)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to list vpc vsi profiles")
+		return nil, errors.Wrapf(err, "failed to list vpc vsi profiles using: %s", c.vpcAPI.Service.Options.URL)
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go:519]: I'll migrate to `fmt.Errorf` given `errors.Wrap` is deprecated (if I recall correctly).

File: data/data/ibmcloud/variables-ibmcloud.tf [hunk 16]
```diff
@@ -110,11 +115,10 @@ variable "ibmcloud_publish_strategy" {
...
   default = "External"
-  # TODO: Supported on tf 0.13
-  # validation {
-  #   condition     = "External" || "Internal"
-  #   error_message = "The ibmcloud_publish_strategy value must be \"External\" or \"Internal\"."
-  # }
+  validation {
+    condition = can(regex("^External$|^Internal$", var.ibmcloud_publish_strategy))
+    error_message = "The ibmcloud_publish_strategy value must be \"External\" or \"Internal\"."
+  }
```
[the remark this answers is not in the document]
Comment (cjschaef) [reply data/data/ibmcloud/variables-ibmcloud.tf:121]: Yes, that should already be validated since those values are enums
https://github.com/openshift/installer/blob/1bacadd1a8853d56d78e397fbc8e40303283cf47/pkg/types/installconfig.go#L65-L74

If we don't believe this is necessary at the Terraform level, I'll drop this validation check here.

[the remark this answers is not in the document]
Comment (cjschaef) [reply data/data/ibmcloud/variables-ibmcloud.tf]: Removed, relying on installconfig/validation

File: pkg/asset/installconfig/ibmcloud/client.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/client.go]: Updated

File: pkg/destroy/bootstrap/bootstrap.go (hunk 11 above)
Comment (cjschaef) [reply pkg/destroy/bootstrap/bootstrap.go]: Updated

File: pkg/destroy/bootstrap/bootstrap.go (hunk 12 above)
Comment (cjschaef) [reply pkg/destroy/bootstrap/bootstrap.go]: Updated

File: pkg/terraform/stages/ibmcloud/stages.go (hunk 13 above)
Comment (cjschaef) [reply pkg/terraform/stages/ibmcloud/stages.go]: Updated

File: pkg/tfvars/ibmcloud/ibmcloud.go (hunk 14 above)
Comment (cjschaef) [reply pkg/tfvars/ibmcloud/ibmcloud.go]: Updated

File: data/data/ibmcloud/variables-ibmcloud.tf (hunk 16 above)
[the remark this answers is not in the document]
Comment (r4f4) [reply data/data/ibmcloud/variables-ibmcloud.tf]: Not a big deal but I'd say not needed. User input validations should be performed before the terraform stage (in fact, as early as possible).

**Comment by r4f4** [comment]
/hold cancel
`openstack-manifests` fix has merged.

**Comment by cjschaef** [comment]
I've rebased and pulled in a new `openshift/api` release for the IBMCloudServiceName enum needed.

File: pkg/asset/installconfig/ibmcloud/validation.go [hunk 17]
```diff
@@ -414,6 +417,52 @@ func ValidatePreExistingPublicDNS(client API, ic *types.InstallConfig, metadata
...
+	if endpoint == "e2e.unittest.local" {
+		return nil
+	}
+	// NOTE(cjschaef): At this time we expect the endpoint to be an absolute URI (besides local unittests checked above)
+	_, err := url.Parse(endpoint)
+	if err != nil {
+		return err
+	}
+	// Verify the endpoint is accessible, the endpoint check should be safe, as we only wish to validate the server responds
+	_, err = http.Head(endpoint) //nolint:gosec
```
Comment (r4f4) [inline pkg/asset/installconfig/ibmcloud/validation.go:462]: It would be nice to add a comment explaining why it's ok to silence the linter
```suggestion
	_, err = http.Head(endpoint) //nolint:gosec // reason goes here
```

File: data/data/ibmcloud/variables-ibmcloud.tf [hunk 18]
```diff
@@ -70,7 +75,7 @@ variable "ibmcloud_vpc_permitted" {
 variable "ibmcloud_vpc" {
   type        = string
   description = "The name of an existing cluster VPC."
-  default     = null
+  default     = ""
```
Comment (r4f4) [inline data/data/ibmcloud/variables-ibmcloud.tf:78]: Is this change related to this PR?

**Review by r4f4 — COMMENTED** [review]
We usually prefer that vendoring changes are in their own commit. It also helps with code reviews, since we can skip them when looking at the commits until all the other parts have been reviewed.

I really wish this whole feature was not in a single commit but broken down into multiple logical units. For examples, 1 commit for terraform configs changes, 1 commit for adding a new field to install-config, 1 commit for replacing existing `errors.Wrap` by `fmt.Errorf`, etc. It would make reviewing much easier.

File: data/data/ibmcloud/variables-ibmcloud.tf (hunk 18 above)
Comment (cjschaef) [reply data/data/ibmcloud/variables-ibmcloud.tf:78]: No, I was cleaning up the reference as I found `null` isn't desired for string.

File: pkg/asset/installconfig/ibmcloud/validation.go (hunk 17 above)
Comment (cjschaef) [reply pkg/asset/installconfig/ibmcloud/validation.go:462]: Sure, I'll add.

**Comment by cjschaef** [comment]
I can attempt to break the PR up into multiple PR's if that helps then.

I would open separate PR's then from this PR, as not to change this one until I know those are valid/expected.

**Comment by r4f4** [comment]
> I can attempt to break the PR up into multiple PR's if that helps then.
> 
> I would open separate PR's then from this PR, as not to change this one until I know those are valid/expected.

That would be nice but if you're aiming to get this merged before FF, just breaking this PR down into commits is enough. Unless you think some part of this work can be merged independently, in which case a separate PR might help to push those along.

**Comment by cjschaef** [comment]
Okay, I went with that, and updated this PR with those changes (3 commits, without the other cleanup changes).

I've opened a separate draft  PR for those cleanup related changes.
https://github.com/openshift/installer/pull/7766

**Comment by patrickdillon** [comment]
This LGTM
looks like there is a small linter error

**Comment by cjschaef** [comment]
I'll see about rebasing and rebuilding the go mod conflicts.

**Review by r4f4 — COMMENTED** [review]
/lgtm
